relay/relay.go

```diff
@@ -263,261 +242,180 @@ func (r *Relay) process(parentCtx context.Context) error {
-				// ref: https://dev.mysql.com/doc/internals/en/binlog-event-flag.html
-				log.Warnf("[relay] skip artificial event %+v", e.Header)
-				continue
-			}
 		}
-
-		if gapReader == nil {
-			gapDetected, fSize, err := r.detectGap(e)
-			if err != nil {
-				relayLogWriteErrorCounter.Inc()
-				return errors.Annotatef(err, "detect relay log file gap for event %+v", e.Header)
-			}
-			if gapDetected {
-				gapSyncStartPos = &mysql.Position{
-					Name: lastPos.Name,
-					Pos:  fSize,
-				}
-				gapSyncEndPos = &mysql.Position{
-					Name: lastPos.Name,
-					Pos:  e.Header.LogPos,
-				}
-				log.Infof("[relay] gap detected from %d to %d in %s, current event %+v", fSize, e.Header.LogPos-e.Header.EventSize, lastPos.Name, e.Header)
-				continue // skip this event after the gap, it will be wrote to the file when filling the gap
-			}
-		} else {
-			// why check gapSyncEndPos != nil?
-			if gapSyncEndPos != nil && e.Header.LogPos >= gapSyncEndPos.Pos {
-				// catch up, after write this event, gap will be filled
-				log.Infof("[relay] fill gap reaching the end pos %v, current event %+v", gapSyncEndPos.String(), e.Header)
-				closeGapSyncer()
-			} else if addRelayFlag {
-				// add LOG_EVENT_RELAY_LOG_F flag to events which used to fill the gap
-				// ref: https://dev.mysql.com/doc/internals/en/binlog-event-flag.html
-				r.addFlagToEvent(e, 0x0040, eventFormat)
-			}
+		if tResult.Ignore {
+			log.Infof("[relay] ignore event %+v by transformer", e.Header)
+			continue
 		}
 
-		cmp, fSize, err := r.compareEventWithFileSize(e)
+		// 3. save events into file
+		writeTimer := time.Now()
+		log.Debugf("[relay] writing binlog event with header %+v", e.Header)
+		wResult, err := writer2.WriteEvent(e)
 		if err != nil {
 			relayLogWriteErrorCounter.Inc()
 			return errors.Trace(err)
-		}
-		if cmp < 0 {
-			log.Warnf("[relay] skip obsolete event %+v (with relay file size %d)", e.Header, fSize)
+		} else if wResult.Ignore {
+			log.Infof("[relay] ignore event %+v by writer", e.Header)
 			continue
-		} else if cmp > 0 {
-			relayLogWriteErrorCounter.Inc()
-			return errors.Errorf("some events missing, current event %+v, lastPos %v, current GTID %v, relay file size %d", e.Header, lastPos, lastGTID, fSize)
 		}
+		relayLogWriteDurationHistogram.Observe(time.Since(writeTimer).Seconds())
 
+		// 4. update meta and metrics
+		needSavePos := tResult.CanSaveGTID
+		lastPos.Pos = tResult.LogPos
+		err = lastGTID.Set(tResult.GTIDSet)
+		if err != nil {
+			log.Errorf("[relay] update last GTID set to %v error %v", tResult.GTIDSet, err)
+		}
 		if !r.cfg.EnableGTID {
 			// if go-mysql set RawModeEnabled to true
 			// then it will only parse FormatDescriptionEvent and RotateEvent
 			// then check `e.Event.(type)` for `QueryEvent` and `XIDEvent` will never be true
 			// so we need to update pos for all events
 			// and also save pos for all events
-			lastPos.Pos = e.Header.LogPos
+			if e.Header.EventType != replication.ROTATE_EVENT {
```

Comment:
we can't swith replication mode, right? like from position to gtid replication

Reply:
can't switch replication mode online now, but can switch after restarted.